File: JiaRemoteWin/JiaRemoteWin/JiaProtocol.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Runtime.InteropServices;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace JiaRemoteWin
{
    public static class JiaProtocol
    {
        public const int DefaultPort = 9527;
        public const int MaxFrameSize = 256 * 1024 * 1024;

        private static readonly byte[] FrameMagic = { 0x4A, 0x52, 0x4D, 0x43 };
        private static readonly byte[] CmdMagic = { 0x4A, 0x52, 0x43, 0x4D };

        public struct FrameHeader
        {
            public uint Width;
            public uint Height;
            public uint BytesPerRow;
            public uint PixelFormat;
            public ulong Timestamp;
            public ulong DataLength;

            public const int HeaderSize = 36;

            public byte[] ToBytes()
            {
                byte[] data = new byte[36];
                Array.Copy(FrameMagic, 0, data, 0, 4);
                BitConverter.GetBytes(Width).CopyTo(data, 4);
                BitConverter.GetBytes(Height).CopyTo(data, 8);
                BitConverter.GetBytes(BytesPerRow).CopyTo(data, 12);
                BitConverter.GetBytes(PixelFormat).CopyTo(data, 16);
                BitConverter.GetBytes(Timestamp).CopyTo(data, 20);
                BitConverter.GetBytes(DataLength).CopyTo(data, 28);
                return data;
            }

            public static FrameHeader? FromBytes(byte[] data)
            {
                if (data.Length < 36) return null;
                if (data[0] != FrameMagic[0] || data[1] != FrameMagic[1] ||
                    data[2] != FrameMagic[2] || data[3] != FrameMagic[3]) return null;
                return new FrameHeader
                {
                    Width = BitConverter.ToUInt32(data, 4),
                    Height = BitConverter.ToUInt32(data, 8),
                    BytesPerRow = BitConverter.ToUInt32(data, 12),
                    PixelFormat = BitConverter.ToUInt32(data, 16),
                    Timestamp = BitConverter.ToUInt64(data, 20),
                    DataLength = BitConverter.ToUInt64(data, 28)
                };
            }
        }

        public enum CommandType : uint
        {
            MouseMove = 1, MouseDown = 2, MouseUp = 3,
            MouseScroll = 4, MouseDblClick = 5,
            KeyDown = 10, KeyUp = 11, KeyCombo = 12,
            SystemCommand = 20,
            ClipboardPush = 30, ClipboardPull = 31,
            WindowListRequest = 40, WindowListResponse = 41,
            WindowFocus = 42, WindowClose = 43,
            DisplayInfoRequest = 50, DisplayInfoResponse = 51,
            Ping = 100, Pong = 101
        }

        public struct MessageHeader
        {
            public CommandType Type;
            public ulong PayloadLength;

            public const int HeaderSize = 16;

            public byte[] ToBytes()
            {
                byte[] data = new byte[16];
                Array.Copy(CmdMagic, 0, data, 0, 4);
                BitConverter.GetBytes((uint)Type).CopyTo(data, 4);
                BitConverter.GetBytes(PayloadLength).CopyTo(data, 8);
                return data;
            }

            public static MessageHeader? FromBytes(byte[] data)
            {
                if (data.Length < 16) return null;
                if (data[0] != CmdMagic[0] || data[1] != CmdMagic[1] ||
                    data[2] != CmdMagic[2] || data[3] != CmdMagic[3]) return null;
                return new MessageHeader
                {
                    Type = (CommandType)BitConverter.ToUInt32(data, 4),
                    PayloadLength = BitConverter.ToUInt64(data, 8)
                };
            }
        }

        public static byte[] EncodeCommand<T>(CommandType type, T payload)
        {
            string json = JsonSerializer.Serialize(payload);
            byte[] payloadBytes = Encoding.UTF8.GetBytes(json);
            byte[] header = new MessageHeader { Type = type, PayloadLength = (ulong)payloadBytes.Length }.ToBytes();
            byte[] result = new byte[header.Length + payloadBytes.Length];
            Array.Copy(header, 0, result, 0, header.Length);
            Array.Copy(payloadBytes, 0, result, header.Length, payloadBytes.Length);
            return result;
        }

        public static T DecodePayload<T>(byte[] payload)
        {
            try { return JsonSerializer.Deserialize<T>(payload); }
            catch { return default; }
        }

        public static byte[] EncodeFrame(byte[] pixelData, uint width, uint height, uint bytesPerRow, uint pixelFormat)
        {
            var header = new FrameHeader
            {
                Width = width, Height = height,
                BytesPerRow = bytesPerRow, PixelFormat = pixelFormat,
                Timestamp = (ulong)DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
                DataLength = (ulong)pixelData.Length
            };
            byte[] headerBytes = header.ToBytes();
            byte[] result = new byte[headerBytes.Length + pixelData.Length];
            Array.Copy(headerBytes, result, headerBytes.Length);
            Array.Copy(pixelData, 0, result, headerBytes.Length, pixelData.Length);
            return result;
        }
    }

    public class MousePoint { public float X { get; set; } public float Y { get; set; } }
    public class MouseButtonEvent { public int Button { get; set; } public MousePoint Point { get; set; } = new(); }
    public class MouseScrollEvent { public float DeltaY { get; set; } public float DeltaX { get; set; } public MousePoint Point { get; set; } = new(); }
    public class KeyEvent { public ushort KeyCode { get; set; } public ulong Flags { get; set; } }
    public class KeyComboEvent { public ushort[] KeyCodes { get; set; } = Array.Empty<ushort>(); public ulong Flags { get; set; } }
    public class SystemCommandEvent { public string CommandType { get; set; } = ""; }
    public class ClipboardData { public string Text { get; set; } = ""; }
    public class RemoteWindowInfo { public uint Id { get; set; } public string Title { get; set; } = ""; public string AppName { get; set; } = ""; public bool IsOnScreen { get; set; } }
    public class RemoteDisplayInfo { public uint Id { get; set; } public ushort Width { get; set; } public ushort Height { get; set; } public ushort RefreshRate { get; set; } }

    public class ConnectionStateChangedEventArgs : EventArgs
    {
        public bool Connected { get; set; }
        public string Host { get; set; } = "";
    }

    public class FrameReceivedEventArgs : EventArgs
    {
        public JiaProtocol.FrameHeader Header { get; set; }
        public byte[] PixelData { get; set; } = Array.Empty<byte>();
    }

    public class CommandReceivedEventArgs : EventArgs
    {
        public JiaProtocol.CommandType Type { get; set; }
        public byte[] Payload { get; set; } = Array.Empty<byte>();
    }
}

File: JiaRemoteMac/Sources/Protocol/JiaProtocol.swift
import Foundation

enum JiaProtocol {

    struct FrameHeader {
        let width: UInt32
        let height: UInt32
        let bytesPerRow: UInt32
        let pixelFormat: UInt32
        let timestamp: UInt64
        let dataLength: UInt64

        static let magicBytes: [UInt8] = [0x4A, 0x52, 0x4D, 0x43]
        static let headerSize = 36

        var headerData: Data {
            var data = Data(capacity: FrameHeader.headerSize)
            data.append(contentsOf: FrameHeader.magicBytes)
            data.append(packUInt32(width))
            data.append(packUInt32(height))
            data.append(packUInt32(bytesPerRow))
            data.append(packUInt32(pixelFormat))
            data.append(packUInt64(timestamp))
            data.append(packUInt64(dataLength))
            return data
        }

        var totalFrameSize: Int {
            FrameHeader.headerSize + Int(dataLength)
        }

        static func decode(from data: Data) -> FrameHeader? {
            guard data.count >= headerSize else { return nil }
            guard data[0] == magicBytes[0],
                  data[1] == magicBytes[1],
                  data[2] == magicBytes[2],
                  data[3] == magicBytes[3] else { return nil }

            var offset = 4
            let width = unpackUInt32(data, offset: offset); offset += 4
            let height = unpackUInt32(data, offset: offset); offset += 4
            let bytesPerRow = unpackUInt32(data, offset: offset); offset += 4
            let pixelFormat = unpackUInt32(data, offset: offset); offset += 4
            let timestamp = unpackUInt64(data, offset: offset); offset += 8
            let dataLength = unpackUInt64(data, offset: offset)

            return FrameHeader(
                width: width,
                height: height,
                bytesPerRow: bytesPerRow,
                pixelFormat: pixelFormat,
                timestamp: timestamp,
                dataLength: dataLength
            )
        }

        func encodeFrame(pixelData: Data) -> Data {
            var frame = Data(capacity: totalFrameSize)
            frame.append(headerData)
            frame.append(pixelData)
            return frame
        }
    }

    enum CommandType: UInt32 {
        case mouseMove = 1
        case mouseDown = 2
        case mouseUp = 3
        case mouseScroll = 4
        case mouseDblClick = 5

        case keyDown = 10
        case keyUp = 11
        case keyCombo = 12

        case systemCommand = 20

        case clipboardPush = 30
        case clipboardPull = 31

        case windowListRequest = 40
        case windowListResponse = 41
        case windowFocus = 42
        case windowClose = 43

        case displayInfoRequest = 50
        case displayInfoResponse = 51

        case ping = 100
        case pong = 101
    }

    struct MessageHeader {
        let commandType: CommandType
        let payloadLength: UInt64

        static let magicBytes: [UInt8] = [0x4A, 0x52, 0x43, 0x4D]
        static let headerSize = 16

        func encode() -> Data {
            var data = Data(capacity: MessageHeader.headerSize)
            data.append(contentsOf: MessageHeader.magicBytes)
            data.append(packUInt32(commandType.rawValue))
            data.append(packUInt64(payloadLength))
            return data
        }

        static func decode(from data: Data) -> MessageHeader? {
            guard data.count >= headerSize else { return nil }
            guard data[0] == magicBytes[0],
                  data[1] == magicBytes[1],
                  data[2] == magicBytes[2],
                  data[3] == magicBytes[3] else { return nil }

            let rawType = unpackUInt32(data, offset: 4)
            let payloadLength = unpackUInt64(data, offset: 8)

            guard let commandType = CommandType(rawValue: rawType) else { return nil }

            return MessageHeader(commandType: commandType, payloadLength: payloadLength)
        }
    }

    struct MousePoint: Codable {
        let x: Float
        let y: Float
    }

    struct MouseButtonEvent: Codable {
        let button: Int
        let point: MousePoint
    }

    struct MouseScrollEvent: Codable {
        let deltaY: Float
        let deltaX: Float
        let point: MousePoint
    }

    struct KeyEvent: Codable {
        let keyCode: UInt16
        let flags: UInt64
    }

    struct KeyComboEvent: Codable {
        let keyCodes: [UInt16]
        let flags: UInt64
    }

    enum SystemCommandType: String, Codable {
        case sleep
        case restart
        case shutdown
        case lockScreen
        case volumeUp
        case volumeDown
        case mute
        case brightnessUp
        case brightnessDown
        case launchpad
        case missionControl
        case wake
    }

    struct SystemCommandEvent: Codable {
        let commandType: String
        let value: Float?
    }

    struct ClipboardData: Codable {
        let text: String
    }

    struct RemoteWindowInfo: Codable {
        let id: UInt32
        let title: String
        let appName: String
        let isOnScreen: Bool
    }

    struct RemoteDisplayInfo: Codable {
        let id: UInt32
        let width: UInt16
        let height: UInt16
        let refreshRate: UInt16
    }

    enum ProtocolConstants {
        static let defaultPort: UInt16 = 9527
        static let frameChannelPrefix = "JR_FRAME"
        static let commandChannelPrefix = "JR_CMD"
        static let maxFrameSize = 256 * 1024 * 1024
        static let compressionFlag: UInt32 = 0x80000000
    }

    static func encodeCommand(type: CommandType, payload: some Encodable) -> Data? {
        guard let payloadData = try? JSONEncoder().encode(payload) else { return nil }

        let header = MessageHeader(commandType: type, payloadLength: UInt64(payloadData.count))
        var command = Data(capacity: MessageHeader.headerSize + payloadData.count)
        command.append(header.encode())
        command.append(payloadData)
        return command
    }

    static func decodeCommandPayload<T: Decodable>(_ type: T.Type, from payload: Data) -> T? {
        try? JSONDecoder().decode(type, from: payload)
    }
}

private func packUInt32(_ value: UInt32) -> Data {
    var v = value.littleEndian
    return Data(bytes: &v, count: 4)
}

private func packUInt64(_ value: UInt64) -> Data {
    var v = value.littleEndian
    return Data(bytes: &v, count: 8)
}

private func unpackUInt32(_ data: Data, offset: Int) -> UInt32 {
    var value: UInt32 = 0
    _ = withUnsafeMutableBytes(of: &value) { ptr in
        data.copyBytes(to: ptr, from: offset ..< offset + 4)
    }
    return UInt32(littleEndian: value)
}

private func unpackUInt64(_ data: Data, offset: Int) -> UInt64 {
    var value: UInt64 = 0
    _ = withUnsafeMutableBytes(of: &value) { ptr in
        data.copyBytes(to: ptr, from: offset ..< offset + 8)
    }
    return UInt64(littleEndian: value)
}
